app/src/main/java/tech/relaycorp/ping/ui/peers/AddPublicPeerViewModel.kt

```diff
@@ -56,50 +55,69 @@ class AddPublicPeerViewModel
 
     init {
-        val form = alias.combine(certificatePicks, ::Form)
-
-        form
-            .onEach { _saveEnabled.value = it.isValid() }
-            .launchIn(backgroundScope)
+        val form = address.combine(certificatePicks, ::Form)
 
         certificatePicks
-            .onEach { _hasCertificate.value = it.isPresent }
+            .element()
+            .onEach { _certificate.value = Optional.of(readFile.getFileName(it)) }
             .launchIn(backgroundScope)
 
         removeCertificateClicks
             .asFlow()
             .onEach {
                 certificatePicks.value = Optional.empty()
-                _hasCertificate.value = false
+                _certificate.value = Optional.empty()
             }
             .launchIn(backgroundScope)
 
         saveClicks
             .asFlow()
             .flatMapLatest { form.take(1) }
-            .filter { it.isValid() }
-            .onEach { f ->
-                val cert = readFile.read(f.certificateUri.get())
-                addPublicPeer.add(f.alias, cert)
-                _finish.finish()
+            .map { f ->
+                if (f.isValid()) {
+                    val cert = readFile.read(f.certificateUri.get())
+                    addPublicPeer.add(f.address, cert)
+                    _finish.finish()
+                } else {
+                    _errors.send(
+                        when {
+                            !f.isAddressValid() -> Error.InvalidAddress
+                            !f.isCertificatePresent() -> Error.MissingCertificate
+                            else -> Error.GenericSave
+                        }
+                    )
+                }
             }
-            .catch {
-                _errors.send(Error.Save)
+            .catch { exception ->
+                _errors.send(
+                    when (exception) {
+                        is InvalidIdentityCertificate -> Error.InvalidCertificate
+                        else -> Error.GenericSave
+                    }
+                )
+                emit(Unit)
             }
             .launchIn(backgroundScope)
     }
 
     enum class Error {
-        Save
+        InvalidAddress, MissingCertificate, InvalidCertificate, GenericSave
     }
 
     data class Form(
-        val alias: String,
+        val address: String,
         val certificateUri: Optional<Uri>
     ) {
-        fun isValid() =
-            alias.isNotBlank()
-                    && alias.matches(Regex("(\\w+\\.)+\\w+"))
-                    && certificateUri.isPresent
+        fun isAddressValid() =
+            address.isNotBlank() && address.matches(PUBLIC_ADDRESS_REGEX)
+
+        fun isCertificatePresent() = certificateUri.isPresent
+
+        fun isValid() = isAddressValid() && isCertificatePresent()
+    }
+
+    companion object {
+        internal val PUBLIC_ADDRESS_REGEX =
+            Regex("([a-zA-Z0-9][a-zA-Z0-9-]{1,61}[a-zA-Z0-9]\\.)+[a-zA-Z]{2,}")
```

Comment:
Did you consider https://developer.android.com/reference/android/util/Patterns#DOMAIN_NAME? I think that'd be more reliable, and it's what we used in the gateway migration screen in the private gateway.

Reply:
Switched. But do note it also includes IP addresses.

Reply:
Sorry, I had no idea that was the case. Can you please revert it?

Reply:
Reverted